ffi/derive/src/convert.rs

```diff
@@ -518,136 +838,152 @@ fn gen_discriminants(
+                    visit_field(&mut ptr_visitor, field);
+                }
             }
         }
-        fn visit_type_ptr(&mut self, node: &syn::TypePtr) {
-            abort!(node, RAW_POINTER_FOUND_MSG);
+        FfiTypeData::Struct(fields) => {
+            for field in fields.iter() {
+                visit_field(&mut ptr_visitor, field);
+            }
         }
     }
-
-    let mut ptr_visitor = PtrVistor;
-    ptr_visitor.visit_data(data);
 }
 
-fn is_non_local(attrs: &[syn::Attribute]) -> bool {
-    let local = parse_quote! {#[ffi_type(local)]};
-    !attrs.iter().any(|attr| *attr == local)
-}
+fn get_enum_repr_type(
+    emitter: &mut Emitter,
+    enum_name: &Ident,
+    repr: &Repr,
+    is_empty: bool,
+) -> syn2::Type {
+    let Some(kind) = repr.kind else {
+        // empty enums are not allowed to have a `#[repr]` attribute
+        // it's an error to use an `#[derive(FfiType)]` on them
+        // but we still want to generate a reasonable error message, so we check for it here
+        if !is_empty {
+            emit!(emitter, enum_name, "Enum representation not specified. Try adding `#[repr(u32)]` or similar");
+        }
+        return syn2::parse_quote! {u32}
+    };
 
-fn is_fieldless_enum(item: &DataEnum) -> bool {
-    item.variants
-        .iter()
-        .all(|variant| matches!(variant.fields, syn::Fields::Unit))
-}
-
-fn assert_is_repr_c(attrs: &[syn::Attribute]) {
-    let repr = find_attr(attrs, "repr");
-    assert!(is_repr_attr(&repr, "C"));
-}
-
-fn enum_size(enum_name: &Ident, repr: &[syn::NestedMeta]) -> Type {
-    if is_repr_attr(repr, "u8") {
-        parse_quote! {u8}
-    } else if is_repr_attr(repr, "i8") {
-        parse_quote! {i8}
-    } else if is_repr_attr(repr, "u16") {
-        parse_quote! {u16}
-    } else if is_repr_attr(repr, "i16") {
-        parse_quote! {i16}
-    } else if is_repr_attr(repr, "u32") {
-        parse_quote! {u32}
-    } else if is_repr_attr(repr, "i32") {
-        parse_quote! {i32}
-    } else {
-        abort!(enum_name, "Enum representation not supported")
+    let ReprKind::Primitive(primitive) = &*kind else {
+        emit!(emitter, &kind.span(), "Enum should have a primitive representation (like `#[repr(u32)]`)");
+        return syn2::parse_quote! {u32}
+    };
+
+    match primitive {
+        ReprPrimitive::U8 => syn2::parse_quote! {u8},
+        ReprPrimitive::U16 => syn2::parse_quote! {u16},
+        ReprPrimitive::U32 => syn2::parse_quote! {u32},
+        ReprPrimitive::U64 => syn2::parse_quote! {u64},
+        ReprPrimitive::I8 => syn2::parse_quote! {i8},
+        ReprPrimitive::I16 => syn2::parse_quote! {i16},
+        ReprPrimitive::I32 => syn2::parse_quote! {i32},
+
+        _ => {
+            emit!(emitter, &kind.span(), "Enum representation not supported");
+            syn2::parse_quote! {u32}
+        }
     }
 }
 
-fn gen_enum_tag_type(enum_name: &Ident, enum_: &DataEnum) -> TokenStream {
+fn gen_enum_tag_type(
```

Comment:
Is it right that this function returns the actual type that is used for enum's discriminant according to the size of variants?

I am confused because in the reference [it is said](https://doc.rust-lang.org/reference/items/enumerations.html#discriminants):

> Under the [default representation](https://doc.rust-lang.org/reference/type-layout.html#the-default-representation), the discriminant is interpreted as an `isize` value. However, the compiler is allowed to use a smaller type (or another means of distinguishing variants) in its actual memory layout.

So, if this function makes an _assumption_ about the compiler's decision, it might be not a correct behaviour. Or it is not applicable here since the macro generates some other repr?